projects/dqn/src/algorithms.py

```diff
@@ -0,0 +1,175 @@
+    Parts of the code adapted from:
+    https://github.com/transedward/pytorch-dqn/blob/master/dqn_learn.py
+    """
+
+    def __init__(
+        self,
+        observation_size,
+        num_actions,
+        num_steps,
+        device,
+        q_function_network_class=DQNNetwork,
+        exploration_rate_scheduler_class=Scheduler,
+        q_function_optimizer=torch.optim.Adam,
+        discount_rate=0.99,
+        q_function_lr=1e-2,
+        exploration_rate=5e-2,
+        exploration_rate_final=1e-3,
+        target_function_update_rate=1e-2,
+        target_function_update_frequency=100,
+        clip_error=True,
+    ):
+
+        self.device = device
+
+        # Initialize networks
+        input_channels = observation_size[2]
+        self.q_function = q_function_network_class(
+            input_channels=input_channels,
+            num_actions=num_actions
+        )
+        self.q_function.to(self.device)
+        self.target_q_function = q_function_network_class(
+            input_channels=input_channels,
+            num_actions=num_actions
+        )
+        self.target_q_function.to(self.device)
+
+        # Initialize optimizer
+        self.q_function_optimizer = q_function_optimizer(
+            self.q_function.parameters(), lr=q_function_lr
+        )
+
+        # Initialize e-greedy scheduler
+        self.exploration_rate_scheduler = exploration_rate_scheduler_class(
+            start_rate=exploration_rate,
+            end_rate=exploration_rate_final,
+            max_steps=num_steps
+        )
+
+        self.num_actions = num_actions
+        self.clip_error = clip_error
+        self.discount_rate = discount_rate
+        self.target_function_update_rate = target_function_update_rate
+        self.target_function_update_frequency = target_function_update_frequency
+
+    def gradient_step(self, experiences):
+        self.q_function_optimizer.zero_grad()
+        output, loss = self.compute_error(experiences)
+        output.backward(loss)
+        self.q_function_optimizer.zero_grad()
+        return loss.sum().item()
+
+    def state_dict(self):
+        return {
+            "q_function": self.q_function.state_dict(),
+            "target_q_function": self.target_q_function.state_dict(),
+            "q_function_optimizer": self.q_function_optimizer.state_dict(),
+            "exploration_rate_scheduler": self.exploration_rate_scheduler.state_dict(),
+        }
+
+    def load_state_dict(self, state_dict):
+        self.q_function.load_state_dict(state_dict["q_function"])
+        self.target_q_function.load_state_dict(state_dict["target_q_function"])
+        self.q_function_optimizer.load_state_dict(state_dict["q_function_optimizer"])
+        self.exploration_rate_scheduler.load_state_dict(
+            state_dict["exploration_rate_scheduler"])
+
+    def get_next_action(self, last_observation):
+        if np.random.random() < self.exploration_rate_scheduler():
```

Comment:
Could also check here whether we are training or evaluating. We only need to be epsilon greedy during training and can act greedy for evaluation.

Reply:
Good idea. This first implementation only has training though, and the reward accrued in the training is episoded is tracked as a performance metric. I left a reminder in docstring that this should only be used in training and exploration should be removed in evaluation